File: PracticeAssignment2Part1/Q14_FloorInBST.java
package PracticeAssignment2Part1;

public class Q14_FloorInBST {
    class Node{
        int data;
        Node left;
        Node right;
        Node(int data){
            this.data = data;
            left = null;
            right = null;
        }
    }
    /**
     * my approach is to solve it recursivly. I plan to have a helper method that get passed in the root,target and a min(floor) vairabel
     * then i will use binary search and traverse thorugh my tree and keep sotring min in a variable.
     * time o(log n) can be o(n) worst case space o(h) height of tree;
     * 
     */
    public int floorInBst(Node root,int target){
        
        if(root==null){
            return -1; // thorwing exception
        }
            
        return helper(root,target,-1);
               
     
    }
    private int helper(Node root,int target,int min){
        if(root==null){
            return min;
        }
        if(root.data <= target && root.data > min){ // keeps track of min
            min = root.data;
        }
        if(root.data > target){ //if node is grater then target we go left
            return helper(root.left,target,min);
        }else{ // else we go right
            return helper(root.right,target,min);
        }


    }
}

File: homework2/danny_tat/q14_FloorInBST.py
# Question 14: FloorInBST
# Technique: Search Binary Search Tree (BST)
# Time: O(log n) avg, O(n) worst
# Space: O(1)

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def floorInBST(root, target):
    floor_val = None
    cur = root
    while cur:
        if cur.data == target:
            return cur.data
        elif cur.data < target:
            floor_val = cur.data
            cur = cur.right
        else:
            cur = cur.left
    return floor_val


root = Node(10)
root.left = Node(8)
root.left.right = Node(9)
root.right = Node(16)
root.right.left = Node(13)
root.right.right = Node(17)
root.right.right.right = Node(20)

# target=13 → 13
print(floorInBST(root, 13))

# target=15 → 13
print(floorInBST(root, 15))

# time: 36 minutes
